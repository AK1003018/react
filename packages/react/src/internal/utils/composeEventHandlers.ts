type Handler<E> = (event: E) => void

export function composeEventHandlers<E>(...handlers: Array<Handler<E> | null | undefined>) {
  return function handleEvent(event: E) {
    for (const handler of handlers) {
      handler?.(event)
      if ((event as unknown as Event).defaultPrevented) {
        break
      }
    }
  }
}
