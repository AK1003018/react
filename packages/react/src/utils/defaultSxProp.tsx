import type {BetterSystemStyleObject} from '../sx'

export const defaultSxProp: Readonly<BetterSystemStyleObject> = Object.freeze<BetterSystemStyleObject>({})
