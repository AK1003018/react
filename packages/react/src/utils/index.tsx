import isNumeric from './isNumeric'
import useIsomorpyhicLayoutEffect from './useIsomorphicLayoutEffect'

export {isNumeric, useIsomorpyhicLayoutEffect}

export * from './createComponent'
export * from './defaultSxProp'
export * from './deprecate'
export * from './environment'
export * from './form-story-helpers'
export * from './getBreakpointDeclarations'
export * from './invariant'
export * from './layout'
export * from './scroll'
export * from './ssr'
export * from './theme'
export * from './use-force-update'
export * from './warning'

export type * from './types'
export type {ForwardRefComponent, OwnProps, IntrinsicElement} from './polymorphic'
