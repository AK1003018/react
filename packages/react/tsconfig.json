{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./lib",
    "typeRoots": ["../../node_modules/@types", "./node_modules/@types", "./@types"],
    "plugins": [{"name": "typescript-plugin-css-modules"}]
  },
  "include": [
    "@types/**/*.ts",
    "src/**/*.js",
    "src/**/*.ts",
    "src/**/*.tsx",
    "script/**/*.ts"
  ],
  "exclude": [
    "src/utils/story-helpers.tsx",
    "**/*.stories.tsx",
    "**/*.test.tsx",
    "**/*.figma.tsx",
    "script/**/*.ts",
    "src/__tests__/",
    "src/utils/test-*.tsx",
    "src/utils/testing.tsx",
    "e2e/**/*.ts",
    "playwright.config.ts",
  ]
}
