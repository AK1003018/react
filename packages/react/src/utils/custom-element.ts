import React from 'react'
import styled from 'styled-components'
import type {EventName} from '@lit-labs/react'
import {createComponent as create} from '@lit-labs/react'
import sx, {type SxProp} from '../sx'
import type { d } from '@storybook/channels/dist/main-c55d8855'

export interface ElementRender {
  renderShadow(): string
  shadowRootOptions?: {
    shadowrootmode?: 'open' | 'closed'
    delegatesFocus?: boolean
  }
}

function hasShadow(elementClass: object): elementClass is ElementRender {
  return 'renderShadow' in elementClass && typeof elementClass.renderShadow === 'function'
}

type EventNames = Record<string, EventName | string>
const rename = (str: string): string => str[0].toUpperCase() + str.slice(1).replace(/(-\w)/g, s => s[1].toUpperCase())
// eslint-disable-next-line @typescript-eslint/ban-types
export const createComponent = <I extends HTMLElement, E extends EventNames = {}>(
  elementClass: new () => I,
  tagName: string,
  events: E | undefined = undefined,
) => {
  const Output = Object.assign(
    styled(
      create<I, E>({
        tagName,
        elementClass,
        react: React,
        events,
      }),
    )(sx),
    {
      displayName: rename(tagName),
    },
  )

  if (hasShadow(elementClass)) {
    // Generate a key for the template
    const templateKey = Math.random().toString(36).slice(2)
    const divKey = Math.random().toString(36).slice(2)
    const dangerouslySetInnerHTML = {__html: elementClass.renderShadow()}
    // const div = React.createElement('div', {dangerouslySetInnerHTML, key: divKey})
    // const documentFragment = new DocumentFragment()
    // documentFragment.appendChild(div)
    return ({children, ...props}: React.ComponentProps<typeof Output> & SxProp) => {
      const {shadowrootmode = 'open', ...templateProps} = elementClass.shadowRootOptions || {}
      const template = React.createElement('template', {shadowrootmode, key: templateKey, dangerouslySetInnerHTML, ...templateProps})
      // @ts-ignore - Type instantiation is excessively deep and possibly infinite
      return React.createElement(Output, props, [template, ...children])
    }
  }

  return Output
}

export default createComponent
