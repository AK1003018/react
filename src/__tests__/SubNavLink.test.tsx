import React from 'react'
import {SubNav} from '..'
import {render, behavesAsComponent} from '../utils/testing'
import {render as HTMLRender} from '@testing-library/react'
import {axe, toHaveNoViolations} from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('SubNav.Link', () => {
  // This component does respect the `as` prop, but it affects the link inside, rather than the wrapping element,
  // which causes the test to fail.
  behavesAsComponent({Component: SubNav.Link, options: {skipAs: true}})

  it('renders an <li> + <a> by default', () => {
    const result = render(<SubNav.Link />)
    expect(result.type).toEqual('li')
    const firstChild = result.children![0]!
    if (typeof firstChild !== 'object') throw new Error(`unexpected child type: ${typeof firstChild}`)
    expect(firstChild.type).toEqual('a')
  })

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(
      <SubNav.Links>
        <SubNav.Link />
      </SubNav.Links>,
    )
    const results = await axe(container)
    expect(results).toHaveNoViolations()
  })

  it('respects the "selected" prop', () => {
    expect(render(<SubNav.Link selected />)).toMatchSnapshot()
  })
})
