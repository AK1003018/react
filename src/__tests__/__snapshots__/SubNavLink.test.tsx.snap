// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubNav.Link renders consistently 1`] = `
.c1 {
  padding-left: 16px;
  padding-right: 16px;
  font-weight: 500;
  font-size: 14px;
  line-height: 20px;
  min-height: 34px;
  color: #1F2328;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-top: 1px solid #d0d7de;
  border-bottom: 1px solid #d0d7de;
  border-right: 1px solid #d0d7de;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:hover,
.c1:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #f6f8fa;
  -webkit-transition: background-color 0.2s ease;
  transition: background-color 0.2s ease;
}

.c1:hover .SubNav-octicon,
.c1:focus .SubNav-octicon {
  color: #656d76;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:first-of-type .SubNav-link {
  border-top-left-radius: 6px;
  border-bottom-left-radius: 6px;
  border-left: 1px solid #d0d7de;
}

.c0:last-of-type .SubNav-link {
  border-top-right-radius: 6px;
  border-bottom-right-radius: 6px;
}

.c0 .SubNav-link.selected {
  color: #ffffff;
  background-color: #0969da;
  border-color: #0969da;
}

.c0 .SubNav-link.selected .SubNav-octicon {
  color: #ffffff;
}

<li
  className="c0 SubNav-item"
>
  <a
    className="c1 SubNav-item SubNav-link"
  />
</li>
`;

exports[`SubNav.Link respects the "selected" prop 1`] = `
.c1 {
  padding-left: 16px;
  padding-right: 16px;
  font-weight: 500;
  font-size: 14px;
  line-height: 20px;
  min-height: 34px;
  color: #1F2328;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-top: 1px solid #d0d7de;
  border-bottom: 1px solid #d0d7de;
  border-right: 1px solid #d0d7de;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:hover,
.c1:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #f6f8fa;
  -webkit-transition: background-color 0.2s ease;
  transition: background-color 0.2s ease;
}

.c1:hover .SubNav-octicon,
.c1:focus .SubNav-octicon {
  color: #656d76;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:first-of-type .SubNav-link {
  border-top-left-radius: 6px;
  border-bottom-left-radius: 6px;
  border-left: 1px solid #d0d7de;
}

.c0:last-of-type .SubNav-link {
  border-top-right-radius: 6px;
  border-bottom-right-radius: 6px;
}

.c0 .SubNav-link.selected {
  color: #ffffff;
  background-color: #0969da;
  border-color: #0969da;
}

.c0 .SubNav-link.selected .SubNav-octicon {
  color: #ffffff;
}

<li
  className="c0 SubNav-item"
>
  <a
    className="c1 SubNav-item selected SubNav-link"
    selected={true}
  />
</li>
`;
